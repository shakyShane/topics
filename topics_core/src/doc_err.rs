use core::convert::From;
use core::fmt::{Display, Formatter};
use core::option::Option;
use core::option::Option::Some;
use core::result::Result::Ok;
use std::path::PathBuf;

#[derive(Debug, thiserror::Error)]
pub enum DocError {
    #[error(
    "could not read file `{}`\nFull path: {}",
    pb.display(),
    abs.display()
    )]
    PathRead {
        pb: PathBuf,
        abs: PathBuf,
        original: std::io::Error,
    },
    #[error(
        "{}",
        .0
    )]
    SerdeLocationErr(LocationError),
    #[error("{}", .0)]
    Unknown(String),
    #[error("File format not supported: {}", .0.display())]
    NotSupported(PathBuf),
}

impl From<anyhow::Error> for DocError {
    fn from(e: anyhow::Error) -> Self {
        DocError::Unknown(e.to_string())
    }
}

impl From<toml::de::Error> for DocError {
    fn from(e: toml::de::Error) -> Self {
        DocError::Unknown(e.to_string())
    }
}

#[derive(Debug)]
pub struct LocationError {
    pub location: Option<Location>,
    pub input_file: Option<PathBuf>,
    pub input_file_src: String,
    pub description: String,
}

#[derive(Debug)]
pub enum Location {
    LineAndCol {
        line: usize,
        column: usize,
    },
    LineAndColRegion {
        line_start: usize,
        line_end: usize,
        line: usize,
        column: usize,
    },
    Region {
        line_start: usize,
        line_end: usize,
    },
    // Unknown,
}

impl Display for LocationError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let _ = writeln!(f);
        if let Some(location) = &self.location {
            match location {
                Location::LineAndColRegion { line, column, .. }
                | Location::LineAndCol { line, column } => {
                    let _ = writeln!(f, "    msg: {}", self.description);
                    let _ = writeln!(
                        f,
                        "   file: {}",
                        self.input_file
                            .as_ref()
                            .map(|f| f.display().to_string())
                            .unwrap_or_else(|| "None".to_string())
                    );
                    if *line != 0 {
                        let _ = writeln!(f, "   line: {}", line);
                        let _ = writeln!(f, " column: {}", column);
                    }
                }
                Location::Region {
                    line_start,
                    line_end,
                } => {
                    let _ = writeln!(f, "           msg: {}", self.description);
                    let _ = writeln!(
                        f,
                        "          file: {}",
                        self.input_file
                            .as_ref()
                            .map(|f| f.display().to_string())
                            .unwrap_or_else(|| "None".to_string())
                    );
                    let _ = writeln!(f, " between lines: {} & {}", line_start, line_end);
                }
            }
        }
        Ok(())
    }
}
