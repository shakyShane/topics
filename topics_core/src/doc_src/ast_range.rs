pub trait AstRangeImpl {
    fn ast_range(&self) -> AstRange;
}

///
/// An AST (tree-like) range selector
///
/// ```rust
/// # fn main() -> anyhow::Result<()> {
/// # use comrak::Arena;
/// # use topics_core::doc_src::MdElements;
/// let input = "# heading";
/// let arena = Arena::new();
/// let md_elements = MdElements::new(input, &arena);
///
/// // tuple of (vec![0], 1) is path + len
/// let html = md_elements.as_html((vec![0], 1));
/// assert_eq!(html, String::from("<h1>heading</h1>\n"));
/// # Ok(())
/// # }
/// ```
#[derive(Debug, Clone, Default)]
pub struct AstRange {
    pub ast_path: Vec<usize>,
    pub ast_len: usize,
}

impl AstRange {
    pub fn range(path: &[usize], len: usize) -> Self {
        Self {
            ast_path: path.to_vec(),
            ast_len: len,
        }
    }
}

impl AstRangeImpl for AstRange {
    fn ast_range(&self) -> AstRange {
        self.clone()
    }
}

impl AstRangeImpl for (Vec<usize>, usize) {
    fn ast_range(&self) -> AstRange {
        AstRange {
            ast_path: self.0.to_vec(),
            ast_len: self.1,
        }
    }
}

impl AstRangeImpl for (&'_ Vec<usize>, usize) {
    fn ast_range(&self) -> AstRange {
        AstRange {
            ast_path: self.0.to_vec(),
            ast_len: self.1,
        }
    }
}
