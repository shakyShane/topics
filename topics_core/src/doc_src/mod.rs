pub mod doc_src;
pub mod yaml_doc_src;

pub use doc_src::*;
pub use yaml_doc_src::*;
