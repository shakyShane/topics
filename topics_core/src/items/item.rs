use crate::host::HostEntriesCheck;
use crate::items::Topic;
use crate::items::{Command, Instruction};
use crate::items::{DependencyCheck, TaskGroup};
use crate::items::{FileExistsCheck, LineMarker};
use std::str::FromStr;

#[derive(Debug, Clone)]
pub enum Item {
    Command(Command),
    FileExistsCheck(FileExistsCheck),
    DependencyCheck(DependencyCheck),
    Instruction(Instruction),
    HostEntriesCheck(HostEntriesCheck),
    Topic(Topic),
    TaskGroup(TaskGroup),
}

#[derive(Debug, Clone)]
pub enum ItemWrap {
    Named(String),
    Item(Item),
}

impl Item {
    pub fn set_name(&mut self, name: &str) {
        match self {
            Item::Command(cmd) => cmd.name = name.into(),
            Item::FileExistsCheck(fec) => fec.name = name.to_string(),
            Item::DependencyCheck(dc) => dc.name = name.to_string(),
            Item::Instruction(inst) => inst.name = name.into(),
            Item::HostEntriesCheck(hec) => hec.name = name.to_string(),
            Item::Topic(top) => top.name = name.into(),
            Item::TaskGroup(tg) => tg.name = name.to_string(),
        };
    }
    pub fn name(&self) -> String {
        match self {
            Item::Command(cmd) => cmd.name.to_string(),
            Item::FileExistsCheck(fec) => fec.name.clone(),
            Item::DependencyCheck(dc) => dc.name.clone(),
            Item::Instruction(inst) => inst.name.to_string(),
            Item::HostEntriesCheck(hec) => hec.name.clone(),
            Item::Topic(top) => top.name.to_string(),
            Item::TaskGroup(tg) => tg.name.clone(),
        }
    }
    pub fn kind_name(&self) -> String {
        match self {
            Item::Command(_) => "Command",
            Item::FileExistsCheck(_) => "File Exists Check",
            Item::DependencyCheck(_) => "Dependency Check",
            Item::Instruction(_) => "Instruction",
            Item::HostEntriesCheck(_) => "Host Entries Check",
            Item::Topic(_) => "Topic",
            Item::TaskGroup(_) => "Task Group",
        }
        .to_string()
    }
    pub fn set_line_start(&mut self, line_start: usize) {
        match self {
            Item::Instruction(inst) => inst.name.set_line_start(line_start),
            Item::Command(cmd) => cmd.name.set_line_start(line_start),
            Item::Topic(topic) => topic.name.set_line_start(line_start),
            _i => todo!("set line start {}", _i.name()),
        }
    }
}

impl FromStr for Item {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "FileExistsCheck" | "fec" => Ok(Item::FileExistsCheck(Default::default())),
            "Topic" | "topic" => Ok(Item::Topic(Default::default())),
            "TaskGroup" | "tg" | "task-group" => Ok(Item::TaskGroup(Default::default())),
            "Command" | "command" | "cmd" => Ok(Item::Command(Default::default())),
            "Instruction" | "inst" | "instruction" => Ok(Item::Instruction(Default::default())),
            "DependencyCheck" | "Dependency Check" | "dep" | "dep-check" => {
                Ok(Item::DependencyCheck(Default::default()))
            }
            _s => Err(anyhow::anyhow!("Not supported yet: {}", _s)),
        }
    }
}
