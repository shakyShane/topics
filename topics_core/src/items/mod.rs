pub use command::*;
pub use dependency::*;
pub use file_exists::*;
pub use host::*;
pub use instruction::*;
pub use item::*;
pub use line_marker::*;
pub use task_group::*;
pub use topic::*;

pub mod command;
pub mod dependency;
pub mod file_exists;
pub mod host;
pub mod instruction;
pub mod item;
pub mod line_marker;
pub mod task_group;
pub mod topic;
