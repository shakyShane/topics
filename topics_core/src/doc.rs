use crate::doc_err::DocError;
use crate::doc_src::{from_serde_yaml_error, parse_md_str, DocSource, TomlError};
use crate::items::item::Item;
use crate::items::ItemWrap;
use crate::{context::Context, items::Topic};
use std::path::PathBuf;

#[derive(Debug, Default)]
pub struct Doc {
    pub source: DocSource,
    pub items: Vec<ItemTracked>,
    pub errors: Vec<DocError>,
}

#[derive(Debug, Clone)]
pub struct ItemTracked {
    pub item: Item,
    pub src_doc: DocSource,
    pub input_file: Option<PathBuf>,
}

pub type DocResult<T, E = DocError> = core::result::Result<T, E>;

impl Doc {
    pub fn topics(&self) -> Vec<Topic> {
        self.items
            .iter()
            .filter_map(|item| match item {
                ItemTracked {
                    item: Item::Topic(topic),
                    ..
                } => Some(topic.clone()),
                _ => None,
            })
            .collect()
    }
    pub fn topic_names(&self) -> Vec<&str> {
        self.items
            .iter()
            .filter_map(|item| match item {
                ItemTracked {
                    item: Item::Topic(topic),
                    ..
                } => Some(topic.name.as_str()),
                _ => None,
            })
            .collect()
    }
    pub fn topic_by_name(&self, name: &str) -> Option<&Topic> {
        self.items.iter().find_map(|item| match item {
            ItemTracked {
                item: Item::Topic(topic),
                ..
            } => {
                if topic.name == name {
                    Some(topic)
                } else {
                    None
                }
            }
            _ => None,
        })
    }
    pub fn from_path_buf(pb: &PathBuf, ctx: &Context) -> DocResult<Self> {
        let doc_src = match pb.extension() {
            None => todo!("what to handle here?"),
            Some(os_str) => match os_str.to_str() {
                None => todo!("what to handle here?"),
                Some("yaml") | Some("yml") => DocSource::yaml(&pb, ctx)?,
                Some("toml") => DocSource::toml(&pb, ctx)?,
                Some("md") | Some("markdown") => DocSource::md(&pb, ctx)?,
                Some(_other) => return Err(DocError::NotSupported(pb.clone())),
            },
        };
        Self::from_doc_src(&pb, doc_src, &ctx)
    }
    pub fn from_doc_src(_pb: &PathBuf, doc_src: DocSource, _ctx: &Context) -> DocResult<Self> {
        let mut doc = Doc {
            source: doc_src,
            ..Default::default()
        };
        match &doc.source {
            DocSource::Yaml(yaml_doc) => {
                for src in &yaml_doc.doc_src_items.items {
                    let item: Result<Item, DocError> = serde_yaml::from_str(&src.content)
                        .map_err(|err| from_serde_yaml_error(&doc, &src, &err));
                    match item {
                        Err(doc_err) => {
                            doc.errors.push(doc_err);
                        }
                        Ok(item) => {
                            doc.items.push(ItemTracked {
                                item,
                                src_doc: DocSource::Yaml((*yaml_doc).clone()),
                                input_file: yaml_doc.input_file.clone(),
                            });
                        }
                    };
                }
            }
            DocSource::Toml(toml_doc) => {
                let items = one_or_many_toml(&toml_doc.file_content).map_err(|err| TomlError {
                    doc: &doc,
                    toml_err: err,
                })?;
                for item in items {
                    doc.items.push(ItemTracked {
                        item,
                        src_doc: DocSource::Toml((*toml_doc).clone()),
                        input_file: toml_doc.input_file.clone(),
                    });
                }
            }
            DocSource::Md(md_doc) => {
                for src in &md_doc.doc_src_items.items {
                    let items = parse_md_str(&src.content)?;
                    for item in items {
                        if let Item::Topic(topic) = &item {
                            for item_wrap in topic.deps.iter().chain(topic.steps.iter()) {
                                if let ItemWrap::Item(item) = item_wrap {
                                    doc.items.push(ItemTracked {
                                        item: item.clone(),
                                        src_doc: DocSource::Md((*md_doc).clone()),
                                        input_file: md_doc.input_file.clone(),
                                    });
                                }
                            }
                        }
                        doc.items.push(ItemTracked {
                            item,
                            src_doc: DocSource::Md((*md_doc).clone()),
                            input_file: md_doc.input_file.clone(),
                        });
                    }
                }
            }
        }
        Ok(doc)
    }
}

fn one_or_many_toml(input: &str) -> Result<Vec<Item>, toml::de::Error> {
    #[derive(Debug, serde::Deserialize)]
    struct TempItems {
        item: Vec<Item>,
    }
    toml::from_str::<TempItems>(input)
        .or_else(|err| {
            if err
                .to_string()
                .contains("missing field `item` at line 1 column 1")
            {
                toml::from_str::<Item>(input).map(|item| TempItems { item: vec![item] })
            } else {
                Err(err)
            }
        })
        .map(|temp| temp.item)
}

#[test]
fn test_toml() -> anyhow::Result<()> {
    let input = r#"
[[item]]
kind = "Topic"
name = "Run unit tests"
steps = [
    "Install root-level dependencies",
    "Run unit tests command"
]
deps = [
    "install nodejs"
]

[[item]]
kind = "Command"
name = "Run unit tests command"
cwd = "."
command = """
echo hello world!
"""

[[item]]
kind = "Command"
name = "Install root-level dependencies"
cwd = "."
command = """
echo hello world!
"""

[[item]]
kind = "DependencyCheck"
name = "install nodejs"
verify = """
node -v
"""
    "#;
    #[derive(Debug, serde::Deserialize)]
    struct Items {
        item: Vec<Item>,
    }
    let items: Items = toml::from_str(input)?;
    assert_eq!(items.item.len(), 4);
    Ok(())
}

#[test]
fn test_toml_single() {
    let input = r#"
kind = "Command"
cwd = "."
name = "Run unit tests command"
command = """
echo "About to install ${MIN_VERSION}"
"""

env.HELLO = { from = "env vars", key = "minimum_skaffold_version" }

    "#;
    let items = one_or_many_toml(input);
    println!("{:?}", items);
}
