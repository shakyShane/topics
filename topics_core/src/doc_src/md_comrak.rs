use std::fmt;
use std::fmt::{Debug, Formatter};

use comrak::nodes::{Ast, AstNode, NodeCodeBlock, NodeHeading, NodeValue};

use crate::doc_src::ast_range::AstRange;
use crate::doc_src::{parse_inline_kind, MdElements};
use crate::items::{Command, Instruction, Item, ItemWrap, LineMarker};
use comrak::arena_tree::Node;
use std::cell::RefCell;

pub(crate) fn process_node<'a>(node: &'a AstNode<'a>, path: &mut Vec<usize>) -> Vec<Item> {
    let mut kind: Option<Item> = None;
    let mut items: Vec<Item> = vec![];
    let first = node.children().take(1).nth(0);
    // let mut other = node.children().skip(1);

    // process the very first item, expected to be a heading, but maybe not?
    if let Some(node) = first {
        let ast = node.data.borrow();
        match &ast.value {
            NodeValue::Heading(NodeHeading { level: 1, .. }) => {
                let start_line = ast.start_line;
                let t = collect_single_line_text(node);
                let item = parse_inline_kind(&t);
                kind = item;
                if let Some(item) = kind.as_mut() {
                    item.set_line_start(start_line)
                }
            }
            _ => {
                todo!("handle other 'first' elements?")
            }
        }
    }

    if let Some(Item::Instruction(inst)) = kind.as_mut() {
        let Instruction { name: _, ast_range } = inst;
        *ast_range = AstRange::range(&path, node.children().count());
    }

    // todo: probably select many command, for MVP just select the first one seen
    if let Some(Item::Command(cmd)) = kind.as_mut() {
        // find a sibling `code block` that we can use as the 'command'
        let node = node.children().enumerate().find_map(|(index, node)| {
            let d = node.data.borrow();
            match d.value {
                NodeValue::CodeBlock(NodeCodeBlock { fenced: true, .. }) => Some((index, d)),
                _ => None,
            }
        });

        if let Some((index, node)) = node {
            if let NodeValue::CodeBlock(code_block) = &node.value {
                let content = std::str::from_utf8(&code_block.literal).unwrap().trim();
                cmd.with_content(content);
                if !code_block.info.is_empty() {
                    let info = std::str::from_utf8(&code_block.info).unwrap().trim();
                    cmd.with_cli_params(info);
                }
                let mut next_range = path.to_vec();
                next_range.push(index);
                cmd.ast_range = AstRange::range(&next_range, 1)
            }
        }
    }

    if let Some(Item::Topic(topic)) = kind.as_mut() {
        let mut list: Vec<(&'_ Node<RefCell<Ast>>, Option<&'_ Node<RefCell<Ast>>>)> = vec![];
        node.children().enumerate().for_each(|(_index, node)| {
            let d = node.data.borrow();
            match &d.value {
                NodeValue::Heading(NodeHeading { level: 2, .. }) => {
                    let empty_node: Option<&'_ Node<RefCell<Ast>>> = None;
                    list.push((node, empty_node));
                }
                NodeValue::List(_node_list) => {
                    let last = list.last_mut();
                    if let Some(last) = last {
                        if last.1.is_none() {
                            last.1 = Some(node)
                        }
                    }
                }
                _ => {}
            }
        });
        for (heading, maybe_list) in list {
            if let Some(list) = maybe_list {
                let heading_kind = collect_single_line_text(heading);
                let heading_line_start = heading.data.borrow().start_line;
                let _list_data = list.data.borrow();
                for node in list.children() {
                    let d = node.data.borrow();
                    match &d.value {
                        NodeValue::Item(_list) => {
                            for node in node.children() {
                                let named_ref = collect_single_line_text(node);
                                let item_line_start = node.data.borrow();
                                match heading_kind.as_str() {
                                    "Steps" => topic
                                        .add_step_named_ref(named_ref, item_line_start.start_line),
                                    "Dependencies" => topic
                                        .add_dep_named_ref(named_ref, item_line_start.start_line),
                                    _ => {}
                                }
                            }
                        }
                        _ => {}
                    }
                }
            }
        }
    }

    if let Some(kind) = kind {
        items.push(kind)
    }

    items
}

pub(crate) fn to_items(md: &'_ MdElements<'_>) -> Vec<Item> {
    let mut path = vec![0];
    let items = process_node(&md.root, &mut path);

    // for node in root.children() {
    //     let ast = node.data.borrow();
    //
    //     if let Some(Item::Instruction(Instruction { ast, .. })) = kind.as_mut() {
    //         println!("skipping before it's an instruction");
    //         continue;
    //     }
    //
    //     match &ast.value {
    //         NodeValue::Heading(NodeHeading { level: 1, .. }) => {
    //             let t = collect_single_line_text(node);
    //             let item = parse_inline_kind(&t);
    //             kind = item;
    //             // println!("got header.. = {} ", t);
    //             // println!("got item.. = {:?} ", item);
    //         }
    //         NodeValue::Heading(NodeHeading { level: 2, .. }) => {}
    //         NodeValue::Heading(NodeHeading { level: 3, .. }) => {}
    //         NodeValue::Paragraph => {
    //             // println!("++p");
    //             // let ast = node.data.clone().into_inner();
    //             // println!("--p");
    //         }
    //         NodeValue::HtmlBlock(html_block) => {
    //             let html = std::str::from_utf8(&html_block.literal).unwrap();
    //             // println!("++html",);
    //             // println!("\t\t|{}|", html);
    //             // println!("--html",);
    //         }
    //         NodeValue::CodeBlock(cb @ NodeCodeBlock { fenced: true, .. }) => {
    //             let NodeCodeBlock { literal, info, .. } = cb;
    //
    //             // `trim_end` is here because the comrak parser adds a new-line to the
    //             // end of code blocks, which may be spec-compliant, but it's possibly
    //             // something we'd rather not forget about later
    //             let literal = std::str::from_utf8(&literal).unwrap().trim_end();
    //             let info = std::str::from_utf8(&info).unwrap();
    //
    //             // if info.is_empty() {
    //             //     items.push(Element::code_block_without_params(literal));
    //             // } else {
    //             //     items.push(Element::code_block(literal, Some(info)));
    //             // }
    //         }
    //         _ => {}
    //     }
    // }
    //
    // if let Some(item) = kind {
    //     items.push(item);
    // }

    items
}

pub fn debug_ast(asts: &[Ast]) -> impl Debug {
    let _s = String::new();
    struct AstDebug(Vec<Ast>);
    impl Debug for AstDebug {
        fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
            let items = self
                .0
                .iter()
                .map(|ast| {
                    let name = match ast.value {
                        NodeValue::Document => "Document",
                        NodeValue::FrontMatter(_) => "FrontMatter",
                        NodeValue::BlockQuote => "BlockQuote",
                        NodeValue::List(_) => "List",
                        NodeValue::Item(_) => "Item",
                        NodeValue::DescriptionList => "DescriptionList",
                        NodeValue::DescriptionItem(_) => "DescriptionItem",
                        NodeValue::DescriptionTerm => "DescriptionTerm",
                        NodeValue::DescriptionDetails => "DescriptionDetails",
                        NodeValue::CodeBlock(_) => "CodeBlock",
                        NodeValue::HtmlBlock(_) => "HtmlBlock",
                        NodeValue::Paragraph => "Paragraph",
                        NodeValue::Heading(_) => "Heading",
                        NodeValue::ThematicBreak => "ThematicBreak",
                        NodeValue::FootnoteDefinition(_) => "FootnoteDefinition",
                        NodeValue::Table(_) => "Table",
                        NodeValue::TableRow(_) => "TableRow",
                        NodeValue::TableCell => "TableCell",
                        NodeValue::Text(_) => "Text",
                        NodeValue::TaskItem(_) => "TaskItem",
                        NodeValue::SoftBreak => "SoftBreak",
                        NodeValue::LineBreak => "LineBreak",
                        NodeValue::Code(_) => "Code",
                        NodeValue::HtmlInline(_) => "HtmlInline",
                        NodeValue::Emph => "Emph",
                        NodeValue::Strong => "Strong",
                        NodeValue::Strikethrough => "Strikethrough",
                        NodeValue::Superscript => "Superscript",
                        NodeValue::Link(_) => "Link",
                        NodeValue::Image(_) => "Image",
                        NodeValue::FootnoteReference(_) => "FootnoteReference",
                    };
                    format!("start_line: {}: {}", ast.start_line, name)
                })
                .collect::<Vec<String>>();
            f.debug_list().entries(items).finish()
        }
    }
    AstDebug(asts.to_vec())
}

///
/// Single-line text items are identifiers and follow special rules.
///
pub(crate) fn collect_single_line_text<'a>(node: &'a AstNode<'a>) -> String {
    node.children()
        .filter_map(|n| match &n.data.borrow().value {
            NodeValue::Text(t) => Some(std::str::from_utf8(t).unwrap().to_string()),
            // todo, preserve this information?
            NodeValue::Code(t) => Some(std::str::from_utf8(t).unwrap().to_string()),
            _ => None,
        })
        .collect::<Vec<String>>()
        .join("")
}
