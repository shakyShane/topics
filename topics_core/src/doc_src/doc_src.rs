use crate::context::Context;
use crate::doc::DocResult;
use crate::doc_src::{MdDocSource, TomlDocSource, YamlDocSource};

use multi_doc::SingleDoc;
use std::path::PathBuf;

pub trait DocSrcImpl: Sized {
    fn from_path_buf(pb: &PathBuf, ctx: &Context) -> DocResult<Self>;
}

#[derive(Debug, Clone)]
pub enum DocSource {
    Yaml(YamlDocSource),
    Toml(TomlDocSource),
    Md(MdDocSource),
}

impl DocSource {
    pub fn file(&self) -> Option<PathBuf> {
        match self {
            DocSource::Yaml(yaml_doc) => yaml_doc.input_file.clone(),
            DocSource::Toml(toml_doc) => toml_doc.input_file.clone(),
            DocSource::Md(md_doc) => md_doc.input_file.clone(),
        }
    }
    pub fn content(&self) -> &str {
        match self {
            DocSource::Yaml(yaml_doc) => yaml_doc.file_content.as_str(),
            DocSource::Toml(toml_doc) => toml_doc.file_content.as_str(),
            DocSource::Md(md_doc) => md_doc.file_content.as_str(),
        }
    }
    pub fn yaml(pb: &PathBuf, ctx: &Context) -> DocResult<Self> {
        Ok(DocSource::Yaml(YamlDocSource::from_path_buf(&pb, ctx)?))
    }
    pub fn toml(pb: &PathBuf, ctx: &Context) -> DocResult<Self> {
        Ok(DocSource::Toml(TomlDocSource::from_path_buf(&pb, ctx)?))
    }
    pub fn md(pb: &PathBuf, ctx: &Context) -> DocResult<Self> {
        Ok(DocSource::Md(MdDocSource::from_path_buf(&pb, ctx)?))
    }
}

impl Default for DocSource {
    fn default() -> Self {
        Self::Yaml(YamlDocSource::default())
    }
}
