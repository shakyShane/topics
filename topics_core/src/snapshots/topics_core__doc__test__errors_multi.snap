---
source: topics_core/src/doc.rs
expression: doc?.errors

---
[
    SerdeYamlErr(
        LocationError {
            location: Some(
                LineAndCol {
                    line_start: 16,
                    line_end: 22,
                    line: 22,
                    column: 1,
                },
            ),
            input_file: None,
            input_file_src: "---\n\nkind: DependencyCheck\nname: global-node\nverify: node -v\nurl: https://www.nodejs.org\n\n---\n\nkind: DependencyCheck\nname: global-yarn\nverify: yarn -v\nurl: https://yarn.sh/legacy\n\n---\n\nkind: Topic\nname: Run screen shot tests\ndeps:\n  - global-node\n  - global-yarn\nsteps\n",
            description: "while parsing a block mapping, did not find expected key at line 22 column 1",
        },
    ),
]
